imagehash.py

```diff
@@ -71,14 +71,18 @@
 
 """
 
-
 def _binary_array_to_hex(arr):
 	"""
-	internal function to make a hex string out of a binary array.
+	Encode a 2D ndarray of bools as a hex string.
 	"""
-	bit_string = ''.join(str(b) for b in 1 * arr.flatten())
-	width = int(numpy.ceil(len(bit_string)/4))
-	return '{:0>{width}x}'.format(int(bit_string, 2), width=width)
+	hash_int = 0
+	for bit in map(int, numpy.nditer(arr)):
+		hash_int = hash_int << 1 | bit
```

Comment:
How about
`hash_int = sum(2**bit for bit in numpy.nditer(arr.flatten()).astype(int))`
or even
`hash_int = (2**numpy.nditer(arr.flatten()).astype(int))).sum()`

Reply:
I'm afraid I don't see how those options could yield a correct result. If I'm not mistaken, `2**bit` would either yield either a 1 or a 2, which I think is not what you intend.

I suspect you mean to have the digit places in `bit`, which is indeed a good option to explore, however I can't think of a good succinct way to inject those values in there off the top of my head at the moment. I'm also slightly cautious about the use of exponentiation with regards to its effect on performance, as I'm not an expert on Python specifics.

Bit shifting is about as simple and fast as it gets, and although it's not a nice one-liner, I'd say it's very readable in this case.

Reply:
Hmm, you are right, my suggestion is incorrect. I was just thinking that any numpy vectorization should be faster than python iterations. Perhaps something like:
`hash_int = (2**numpy.arange(arr.size).reshape(arr.shape) * arr)).sum()`

Reply:
That was my intuition at first too, however I found out that since numpy uses fixed width integers internally, for any matrix above around 8 it starts overflowing and producing garbage results. The proposed method relies on python's infinite precision ints, and I can't really think of a way to shift it over into C.